File: apps/ios/Exponential/UI/Session/AgentLoginSheet.swift
import ExpCore
import ExpUI
import SwiftUI

/// EXP-862: the sign-in sheet — the ONE place an agent login is driven from a
/// phone, and the mobile twin of the web/desktop login dialog: a title, ONE
/// status line, the machine's own sign-in link, and (for Claude) the field that
/// hands the browser's code back. It closes itself the moment the machine
/// reports the login as signed in.
///
/// It used to live inside the device settings sheet, mixed into that agent's
/// tab beside the launch defaults. EXP-862 split the two jobs: device settings
/// is the machine's configuration, a login is a flow, and the chip menus (a
/// device row's, an account row's "+") open this directly.
///
/// No credential is ever held, copied or forwarded: the machine runs
/// `claude auth login` / `codex login` locally and publishes only the sign-in
/// link it puts on its own screen (EXP-484). EXP-765 closes the loop for
/// Claude, whose browser hands an authorization code back to a CLI still
/// waiting on the machine.
struct AgentLoginTarget: Identifiable {
    let deviceId: String
    let deviceLabel: String
    let agent: String
    /// The EXISTING login to sign in again, `system`/nil for the machine's
    /// ambient one.
    var profileId: String? = nil
    /// EXP-827: create a new profile with this label first and sign into that
    /// one ("+ Add account" on a machine whose ambient login is taken). Never
    /// together with `profileId` — the server refuses that pair.
    var newProfileLabel: String? = nil

    var id: String {
        "\(deviceId):\(agent):\(profileId ?? newProfileLabel ?? "active")"
    }
}

struct AgentLoginSheet: View {
    let viewModel: AgentsViewModel
    let target: AgentLoginTarget

    @Environment(AppDependencies.self) private var deps
    @Environment(\.accountId) private var accountId
    @Environment(\.dismiss) private var dismiss

    /// The queued `agent_login` command id, while one is in flight.
    @State private var pendingLogin = false
    /// The sign-in payload the finished command carried (the machine completes
    /// it EARLY, the moment the link is on its own screen).
    @State private var loginResult: String?
    @State private var errorText: String?
    /// EXP-765: the code typed back from the browser, and what the machine said
    /// about it.
    @State private var codeDraft = ""
    @State private var pendingCode = false
    @State private var codeResult: String?
    @State private var started = false

    var body: some View {
        GlassSheetChrome(title: "Sign in") {
            VStack(alignment: .leading, spacing: 12) {
                statusLine
                link
                if let codeResult {
                    Text(codeResult)
                        .font(.caption)
                        .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                        .fixedSize(horizontal: false, vertical: true)
                }
                if let errorText {
                    Text(errorText)
                        .font(.caption)
                        .foregroundStyle(DesignTokens.Semantic.red)
                        .fixedSize(horizontal: false, vertical: true)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(16)
        }
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("agent-login-sheet")
        .onAppear {
            guard !started else { return }
            started = true
            queueLogin()
        }
        // Closes itself on success: the machine re-probes after the login and
        // its next heartbeat reports the profile as signed in.
        .onChange(of: signedIn) { was, now in
            if !was, now { dismiss() }
        }
    }

    /// The ONE status line (the pinned ×4 shape): who is signing in where, and
    /// what is happening right now — never a paragraph.
    private var statusLine: some View {
        HStack(spacing: 8) {
            if pendingLogin || pendingCode {
                ProgressView().controlSize(.small).tint(.white)
            }
            Text(statusText)
                .font(.caption)
                .foregroundStyle(.white.opacity(TextOpacity.secondary))
                .fixedSize(horizontal: false, vertical: true)
            Spacer(minLength: 0)
        }
    }

    private var statusText: String {
        let agent = LaunchVocabulary.agentLabel(target.agent)
        if pendingCode { return "Sending the code to \(target.deviceLabel)…" }
        if pendingLogin {
            return "Waiting for \(target.deviceLabel) to publish the \(agent) sign-in link…"
        }
        return "\(agent) on \(target.deviceLabel)"
    }

    /// The published link, its code, and Claude's way back.
    @ViewBuilder
    private var link: some View {
        if let result = loginResult {
            if let parsed = AgentUsagePresentation.parseAgentLoginResult(result),
               let url = URL(string: parsed.url) {
                let wantsCodeBack = parsed.code == nil
                VStack(alignment: .leading, spacing: 8) {
                    Link(destination: url) {
                        Label("Open the sign-in link", appIcon: AppIcons.uiExternalLink)
                    }
                    if let code = parsed.code {
                        HStack(spacing: 8) {
                            Text(code)
                                .font(.caption.monospaced())
                            Button {
                                Platform.copyToPasteboard(code)
                            } label: {
                                AppIcon(AppIcons.uiCopy, size: AppIcon.Size.small)
                                    .foregroundStyle(.secondary)
                            }
                            .buttonStyle(.plain)
                            .accessibilityLabel("Copy code")
                        }
                    }
                    if wantsCodeBack {
                        codeEntry
                    }
                    Text(linkCaption(hasCode: parsed.code != nil, wantsCodeBack: wantsCodeBack))
                        .font(.caption)
                        .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                        .fixedSize(horizontal: false, vertical: true)
                }
            } else {
                // Not a link payload (an older build, a plain note): verbatim.
                Text(result)
                    .font(.caption)
                    .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
    }

    /// The caption under a published sign-in link, in the three shapes it
    /// takes: codex's code goes into the browser, Claude's comes back here,
    /// and a machine too old to take it back gets the bare instruction.
    private func linkCaption(hasCode: Bool, wantsCodeBack: Bool) -> String {
        if hasCode {
            return "Open the link on any device and enter the code on the machine."
        }
        return wantsCodeBack
            ? "Open the link on any device, then paste the code it shows here."
            : "Open the link on any device."
    }

    private var codeEntry: some View {
        let typed = codeDraft.trimmingCharacters(in: .whitespacesAndNewlines)
        return HStack(spacing: 8) {
            GlassTextField("Code from the browser", text: $codeDraft)
                .font(.caption.monospaced())
                .textInputAutocapitalization(.never)
                .autocorrectionDisabled()
                .onSubmit { submitCode() }
            GlassPill(
                "Enter code",
                icon: AppIcons.uiSignIn,
                mode: .action { submitCode() },
                enabled: !typed.isEmpty && !pendingCode
            )
        }
    }

    // MARK: - Commands

    /// The machine's live report for the login this sheet is signing in.
    private var signedIn: Bool {
        let device = viewModel.devices?.first { $0.deviceId == target.deviceId }
        guard let account = device?.agentAccounts?[target.agent] else { return false }
        guard let profileId = target.profileId,
              profileId != AgentAccountsRows.systemProfileId,
              let profiles = account.profiles
        else { return account.signedIn == true }
        return profiles.first { $0.id == profileId }?.signedIn == true
    }

    private func queueLogin() {
        errorText = nil
        loginResult = nil
        codeResult = nil
        codeDraft = ""
        pendingLogin = true
        Task {
            let outcome = await run(
                kind: "agent_login",
                profileId: target.newProfileLabel == nil ? target.profileId : nil,
                newProfileLabel: target.newProfileLabel
            )
            pendingLogin = false
            switch outcome {
            case let .done(result): loginResult = result
            case let .failed(message): errorText = message
            case .unanswered: break
            }
        }
    }

    private func submitCode() {
        let code = codeDraft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !code.isEmpty, !pendingCode else { return }
        errorText = nil
        codeResult = nil
        pendingCode = true
        Task {
            let outcome = await run(kind: "agent_login_code", code: code)
            pendingCode = false
            switch outcome {
            case let .done(result):
                // The code is in: the link it answered is spent, so it goes and
                // the machine's own words take its place.
                loginResult = nil
                codeDraft = ""
                codeResult = result
            case let .failed(message): errorText = message
            case .unanswered: break
            }
        }
    }

    private enum CommandOutcome {
        case done(String?)
        case failed(String)
        /// An offline machine keeps the command queued server-side; the poll
        /// simply stops watching and the outcome lands via sync.
        case unanswered
    }

    /// Queue the command, then poll its row ~2s until terminal (bounded).
    private func run(
        kind: String,
        code: String? = nil,
        profileId: String? = nil,
        newProfileLabel: String? = nil
    ) async -> CommandOutcome {
        do {
            let created = try await deps.devicesApi.createCommand(
                accountId: accountId,
                deviceId: target.deviceId,
                kind: kind,
                agent: target.agent,
                code: code,
                profileId: profileId,
                newProfileLabel: newProfileLabel
            )
            for _ in 0..<60 {
                try? await Task.sleep(for: .seconds(2))
                if Task.isCancelled { return .unanswered }
                guard let command = try? await deps.devicesApi.getCommand(
                    accountId: accountId, commandId: created.id
                ) else { continue }
                guard !command.isPending else { continue }
                return command.isFailed
                    ? .failed(command.result ?? "The machine refused the command.")
                    : .done(command.result)
            }
            return .unanswered
        } catch {
            return .failed(error.userFacingMessage)
        }
    }
}

/// EXP-862: "+ Add account" — the Accounts header's pill, and the mobile twin
/// of web's `AddAccountDialog`: pick one of MY online machines that can take a
/// sign-in, pick the agent, and the login runs there. A machine whose ambient
/// login is still free takes it; otherwise the machine creates a NEW profile
/// first (EXP-792's per-agent config dirs) and signs into that one.
struct AddAccountSheet: View {
    let viewModel: AgentsViewModel

    @Environment(\.dismiss) private var dismiss

    @State private var deviceId = ""
    @State private var agent = ""
    @State private var loginTarget: AgentLoginTarget?

    /// The caller's machines a sign-in can be queued on right now: own, online,
    /// advertising `agent-login`, with at least one agent installed.
    private var candidates: [SteerDevice] {
        (viewModel.devices ?? []).filter {
            $0.isMine && $0.isOnline && $0.canAgentLogin && !addableAgents($0).isEmpty
        }
    }

    private var selected: SteerDevice? {
        candidates.first { $0.deviceId == deviceId } ?? candidates.first
    }

    /// Every agent INSTALLED on the machine — runnable or signed out; either
    /// can take a login.
    private func addableAgents(_ device: SteerDevice) -> [String] {
        let set = Set(device.agentIds).union(device.unauthedAgentIds)
        return DomainContract.codingAgentValues.filter { set.contains($0) }
    }

    private var agents: [String] {
        selected.map(addableAgents) ?? []
    }

    private var resolvedAgent: String {
        agents.contains(agent) ? agent : (agents.first ?? "")
    }

    var body: some View {
        GlassSheetChrome(
            title: "Add account",
            content: {
                VStack(spacing: 2) {
                    if candidates.isEmpty {
                        Text("None of your machines can take a sign-in right now. Start the Exponential desktop app or the daemon and try again.")
                            .font(.caption)
                            .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                            .fixedSize(horizontal: false, vertical: true)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding(.horizontal, 12)
                            .padding(.vertical, 12)
                            .glassRow()
                    } else {
                        GlassPickerRow(
                            "Device",
                            selection: Binding(
                                get: { selected?.deviceId ?? "" },
                                set: { deviceId = $0 }
                            ),
                            options: candidates.map(\.deviceId),
                            label: { id in
                                candidates.first { $0.deviceId == id }
                                    .map(LaunchVocabulary.deviceCaption) ?? id
                            },
                            icon: { id in
                                candidates.first { $0.deviceId == id }?.isServer == true
                                    ? AppIcons.uiServer
                                    : AppIcons.uiDevice
                            }
                        )
                        .padding(.horizontal, 12)
                        .padding(.vertical, 12)
                        .glassRow()

                        GlassPickerRow(
                            "Agent",
                            selection: Binding(
                                get: { resolvedAgent },
                                set: { agent = $0 }
                            ),
                            options: agents,
                            label: { LaunchVocabulary.agentLabel($0) }
                        )
                        .padding(.horizontal, 12)
                        .padding(.vertical, 12)
                        .glassRow()
                    }
                }
                .padding(.horizontal, GlassSheetTokens.headerHPadding)
                .padding(.bottom, 16)
            },
            primaryAction: {
                GlassSubmitButton("Sign in", enabled: selected != nil && !resolvedAgent.isEmpty) {
                    guard let device = selected else { return }
                    loginTarget = AgentLoginTarget(
                        deviceId: device.deviceId,
                        deviceLabel: LaunchVocabulary.deviceName(device),
                        agent: resolvedAgent,
                        profileId: ambientProfileId(device, agent: resolvedAgent),
                        newProfileLabel: newProfileLabel(device, agent: resolvedAgent)
                    )
                }
            }
        )
        .accessibilityIdentifier("add-account-sheet")
        .sheet(item: $loginTarget) { target in
            AgentLoginSheet(viewModel: viewModel, target: target)
        }
    }

    /// Where a new login lands: the ambient login while it is signed out
    /// (nothing to keep beside it), otherwise a new profile.
    private func ambientSignedIn(_ device: SteerDevice, agent: String) -> Bool {
        guard let account = device.agentAccounts?[agent] else { return false }
        guard let ambient = account.profiles?.first(where: {
            $0.id == AgentAccountsRows.systemProfileId
        }) else { return account.signedIn == true }
        return ambient.signedIn == true
    }

    private func ambientProfileId(_ device: SteerDevice, agent: String) -> String? {
        ambientSignedIn(device, agent: agent) ? nil : AgentAccountsRows.systemProfileId
    }

    /// `Claude Code account 2` — one past the profiles the machine reports for
    /// the agent (the ambient login counts as the first). Clamped at the
    /// server's 64.
    private func newProfileLabel(_ device: SteerDevice, agent: String) -> String? {
        guard ambientSignedIn(device, agent: agent) else { return nil }
        let profiles = device.agentAccounts?[agent]?.profiles ?? []
        let count = max(profiles.count, 1)
        let label = "\(LaunchVocabulary.agentLabel(agent)) account \(count + 1)"
        return String(label.prefix(64))
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/session/AgentLoginSheet.kt
package com.exponential.app.ui.session

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import com.exponential.app.domain.AgentAccountsRows
import androidx.compose.runtime.remember
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.platform.LocalUriHandler
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.exponential.app.data.api.SteerDevice
import com.exponential.app.domain.parseAgentLoginResult
import com.exponential.app.ui.components.GlassPill
import com.exponential.app.ui.components.GlassSheet
import com.exponential.app.ui.components.GlassTextField
import com.exponential.app.ui.components.agentLabel
import com.exponential.app.ui.icons.ExpIcons
import com.exponential.app.ui.theme.TextEmphasis

/**
 * EXP-862: WHERE a sign-in lands — the machine, the agent, and either an
 * EXISTING login of that agent ([profileId]) or a NEW profile the machine
 * creates first ([newProfileLabel], the "Add account" path). Exactly one of the
 * two is ever set; the ambient login is `system`.
 */
internal data class AgentLoginTarget(
    val device: SteerDevice,
    val agent: String,
    val profileId: String? = null,
    val newProfileLabel: String? = null,
)

/**
 * EXP-862: THE sign-in of this client. A login never travels — the machine runs
 * the agent CLI's OWN flow and publishes its link (plus codex's device code)
 * back as the command result, which is what this sheet renders; claude's flow
 * hands a code back the other way, so the link gets a field and an "Enter code"
 * pill.
 *
 * One implementation for every entry point (a chip's "Sign in", an account's
 * `+`, the Add-account pill) — the device-settings sheet carries no accounts at
 * all since this wave, so there is nowhere else a login can appear.
 */
@Composable
internal fun AgentLoginSheet(
    target: AgentLoginTarget,
    onDismiss: () -> Unit,
    /** The LIVE row of the device the login runs on (the synced heartbeat),
     * so the sheet can see its own success; null keeps it open until
     * dismissed by hand. */
    liveDevice: SteerDevice? = null,
    viewModel: DeviceSettingsViewModel = hiltViewModel(),
) {
    val device = target.device
    val commandStates by viewModel.commandStates.collectAsStateWithLifecycle()

    // EXP-862: the sheet closes itself once the login it asked for is usable
    // on the device (web/desktop/iOS do the same). The FIRST observation is
    // the baseline: a login that was already there never counts as landing.
    val landed = AgentAccountsRows.loginLanded(
        liveDevice?.agentAccounts?.get(target.agent),
        target.profileId,
        target.newProfileLabel,
    )
    var baseline by remember(target) { mutableStateOf<Boolean?>(null) }
    LaunchedEffect(landed) {
        when (baseline) {
            null -> baseline = landed
            false -> if (landed) onDismiss()
            else -> Unit
        }
    }

    // The sheet IS the request: it queues the login once, on the machine the
    // caller named, and then follows it. A second open re-runs it, which is
    // what a stale or abandoned link needs anyway.
    LaunchedEffect(target) {
        device.rowId?.let { viewModel.bind(it) }
        viewModel.agentLogin(
            deviceId = device.deviceId,
            agent = target.agent,
            switchAccount = false,
            deviceOnline = device.online,
            profileId = target.profileId,
            newProfileLabel = target.newProfileLabel,
        )
    }

    val state = commandStates[agentLoginCommandKey(target.agent)]
    val codeState = commandStates[agentLoginCodeCommandKey(target.agent)]
    GlassSheet(
        title = "Sign in",
        onDismiss = onDismiss,
        modifier = Modifier.testTag("agent-login-sheet"),
    ) {
        Column(modifier = Modifier.fillMaxWidth().padding(horizontal = 20.dp)) {
            Text(
                "${agentLabel(target.agent)} signs in on " +
                    "${device.deviceLabel.ifBlank { device.deviceId }}.",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary),
            )
            LoginResultCaption(
                agent = target.agent,
                state = state,
                codeState = codeState,
                // EXP-765: claude's browser hands a code back; the machine has
                // to be listening for it, which is the same gate as the login.
                canEnterCode = device.online && device.isMine,
                onEnterCode = { code ->
                    viewModel.agentLoginCode(device.deviceId, target.agent, code, device.online)
                },
            )
            CommandCaption(codeState)
            Spacer(Modifier.height(16.dp))
        }
    }
}

/**
 * What a queued `agent_login` command is doing (EXP-484). A completed one
 * publishes JSON — the machine's login URL, plus codex's device code — which
 * renders as an openable link and a copyable code; anything else (queued,
 * failed, a result that isn't a login publication) falls through to the
 * ordinary command caption.
 *
 * EXP-765: a link WITHOUT a code is claude's (`code=true` — the browser shows
 * the code and the CLI on the machine is still waiting for it), so when the
 * machine can take it back the link gets a field and an "Enter code" pill; the
 * `agent_login_code` command's own progress captions right below.
 */
@Composable
internal fun LoginResultCaption(
    agent: String,
    state: DeviceCommandUiState?,
    codeState: DeviceCommandUiState?,
    canEnterCode: Boolean,
    onEnterCode: (String) -> Unit,
) {
    val login = (state as? DeviceCommandUiState.Done)?.let { parseAgentLoginResult(it.message) }
    when {
        state is DeviceCommandUiState.Sending || state is DeviceCommandUiState.Running ->
            Text(
                "Waiting for the sign-in link…",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary),
                modifier = Modifier.padding(top = 4.dp),
            )
        login != null -> {
            val uriHandler = LocalUriHandler.current
            val clipboard = LocalClipboardManager.current
            Spacer(Modifier.height(4.dp))
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth().clickable { uriHandler.openUri(login.url) },
            ) {
                Icon(
                    ExpIcons.uiExternalLink,
                    contentDescription = null,
                    modifier = Modifier.size(14.dp),
                    tint = MaterialTheme.colorScheme.primary,
                )
                Spacer(Modifier.width(6.dp))
                Text(
                    login.url,
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.primary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
            login.code?.let { code ->
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        code,
                        style = MaterialTheme.typography.bodyMedium,
                        fontFamily = FontFamily.Monospace,
                        color = MaterialTheme.colorScheme.onSurface,
                    )
                    IconButton(onClick = { clipboard.setText(AnnotatedString(code)) }) {
                        Icon(
                            ExpIcons.uiCopy,
                            contentDescription = "Copy code",
                            modifier = Modifier.size(14.dp),
                            tint = MaterialTheme.colorScheme.onSurface.copy(
                                alpha = TextEmphasis.Tertiary,
                            ),
                        )
                    }
                }
            }
            // Only the codeless (claude) flow returns a code to us; codex's is
            // typed into the browser, so nothing comes back and nothing is
            // offered here.
            val returnsCode = login.code == null && canEnterCode
            Text(
                when {
                    login.code != null ->
                        "Open the link on any device and enter the code on the machine."
                    returnsCode ->
                        "Open the link on any device, then paste the code it shows here."
                    else -> "Open the link on any device."
                },
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
            )
            if (returnsCode) {
                // Per agent: switching sign-ins must not carry a half-typed
                // code over to another agent's.
                key(agent) {
                    var draft by rememberSaveable(agent) { mutableStateOf("") }
                    val sending = codeState is DeviceCommandUiState.Sending ||
                        codeState is DeviceCommandUiState.Running
                    Spacer(Modifier.height(6.dp))
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        GlassTextField(
                            value = draft,
                            onValueChange = { draft = it },
                            placeholder = "Code from the browser",
                            singleLine = true,
                            keyboardOptions = KeyboardOptions(
                                autoCorrectEnabled = false,
                                capitalization = KeyboardCapitalization.None,
                            ),
                            textStyle = MaterialTheme.typography.bodySmall,
                            modifier = Modifier.weight(1f),
                        )
                        GlassPill(
                            "Enter code",
                            onClick = {
                                val code = draft.trim()
                                if (code.isNotEmpty() && !sending) {
                                    onEnterCode(code)
                                    draft = ""
                                }
                            },
                            icon = ExpIcons.uiSignIn,
                            enabled = draft.isNotBlank() && !sending,
                            loading = sending,
                        )
                    }
                }
            }
        }
        else -> CommandCaption(state)
    }
}
